TBG.Business/SingleEliminationTournament.cs

```diff
@@ -9,14 +9,124 @@ namespace TBG.Business
+        {
+            Queue<ITournamentEntry> teamQueue = new Queue<ITournamentEntry>();
+            foreach (var team in Participants)
+            {
+                teamQueue.Enqueue(team);
+            }
+
+            for (int i = 1; i <= CalculateRoundTotal(Participants.Count); i++)
+            {
+                Rounds.Add(new Round()
+                {
+                    RoundNum = i,
+                    TournamentId = TournamentId
+                });
+            }
+
+            foreach (var round in Rounds)
+            {
+                Matchup nextRoundMatchup = new Matchup();
+
+                if (round.RoundNum == 1)
+                {
+                    for (int i = 0; i < Participants.Count / 2; i++)
+                    {
+                        round.Pairings.Add(new Matchup()
+                        {
+                            MatchupId = i,
+                            Teams = new List<IMatchupEntry>()
+                        });
+                    }
+
+                    foreach (var matchup in round.Pairings)
+                    {
+                        matchup.Teams.Add(new MatchupEntry()
+                        {
+                            TheTeam = teamQueue.Dequeue(),
+                            Score = 0
+                        });
+
+                        matchup.Teams.Add(new MatchupEntry()
+                        {
+                            TheTeam = teamQueue.Dequeue(),
+                            Score = 0
+                        });
+                    }
+                }
+
+                if (round.Pairings.Count > 1)
+                {
+                    int count = 1;
+                    foreach (var matchup in round.Pairings)
+                    {
+                        if (IsOdd(count))
+                        {
+                            nextRoundMatchup = new Matchup();
+                            var nextRound = Rounds.Find(x => x.RoundNum == round.RoundNum + 1);
+                            if (nextRound != null) { nextRound.Pairings.Add(nextRoundMatchup); } //Safety, but there should be a next round if Pairings.count > 1.
+                        }
+                        matchup.NextRound = nextRoundMatchup;
+
+                        count++;
+                    }
+                }
+            }
+            return true;
+        }
+
+        public bool RecordResult(IMatchup matchup)
+        {
+            var winner = CreateWinnerMatchupEntry(matchup.Teams.OrderByDescending(x => x.Score).First());
+            matchup.NextRound.Teams.Add(winner);
+            return true;
+        }
+
+        private IMatchupEntry CreateWinnerMatchupEntry(IMatchupEntry winner)
+        {
+            var nextRound = new MatchupEntry()
+            {
+                TheTeam = winner.TheTeam,
+                Score = 0
+            };
+
+            return nextRound;
+        }
+
+        private int CalculateRoundTotal(int FieldSize)
```

Comment:
This function is effectively the same as Log_2 (FieldSize) so we can either put a 1 liner in here, something like return "Math.log(fieldSize,2)" or just remove the function entirely.

Reply:
Yes, we can.